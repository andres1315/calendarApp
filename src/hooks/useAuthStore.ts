import { useEffect } from "react"
import { calendarApi } from "../api/calendarApi"
import { clearError, onLogin, onLogout } from "../store/auth/authSlice"
import { useAppDispatch, useAppSelector } from "../store/hooks"
import Swal from "sweetalert2"
import { set } from "date-fns"
import { onLogoutCalendar } from "../store/calendar/calendarSlice"

export const useAuthStore =()=>{
  const {status, user, errorMessage} = useAppSelector(state=>state.auth)
  const dispatch = useAppDispatch()

  useEffect(()=>{
    if(errorMessage !== '') {
      Swal.fire({
        icon:'error',
        title:'Error en la autenticaciÃ³n',
        text: errorMessage,

      })
    }

  },[errorMessage])

  const startLogin = async({email, password} :{email:string, password:string})=>{
    try{
      const {data} = await calendarApi.post('/auth',{email, password})
      const {token, uid,name} =  data
      localStorage.setItem('token',token)
      dispatch(onLogin({user:{uid,name},token}))

    }catch(error){
      if(error.response.data?.errors){
        const listErrors = Object.values(error.response.data.errors)
        const message = listErrors.map(({msg})=> msg).join(' & ')
        dispatch(onLogout({errorMessage:message}))
        setTimeout(() => {dispatch(clearError())}, 10);

        return
      }
      const message = error.response.data.msg || 'Unexpected error'
      dispatch(onLogout({errorMessage:message}))
      setTimeout(() => {dispatch(clearError())}, 10);
    }
  }

  const startRegister = async({email, password, name} :{email:string, password:string, name:string})=>{
    try{
      const {data} = await calendarApi.post('/auth/new',{email, password, name})
      const {token, uid,name:nameUser} =  data
      localStorage.setItem('token',token)
      dispatch(onLogin({user:{uid,nameUser},token}))

    }catch(error){
      if(error.response.data?.errors){
        const listErrors = Object.values(error.response.data.errors)
        const message = listErrors.map(({msg})=> msg).join(' & ')
        dispatch(onLogout({errorMessage:message}))
        setTimeout(() => {dispatch(clearError())}, 10);
        
        return
      }
      const message = error.response.data.msg || 'Unexpected error'
      dispatch(onLogout({errorMessage:message}))
      setTimeout(() => {dispatch(clearError())}, 10);
    }
  }
  
  const checkToken = async()=>{
    const token = localStorage.getItem('token')
    if(!token) return dispatch(onLogout({errorMessage:''}))
    try{
      const {data} = await calendarApi.get('/auth/renew')
      const {token:tokenNew, uid,name} =  data
      localStorage.setItem('token',tokenNew)
      dispatch(onLogin({user:{uid,name},token:tokenNew}))

    }catch(error){
      localStorage.removeItem('token')
      dispatch(onLogout({errorMessage:''}))
    }
  }

  const startLogout=()=>{
    localStorage.removeItem('token')
    dispatch(onLogoutCalendar())
    dispatch(onLogout({errorMessage:''}))
  }
  return {
    user,
    status,
    errorMessage,
    startLogin,
    startRegister,
    checkToken,
    startLogout
  }
}