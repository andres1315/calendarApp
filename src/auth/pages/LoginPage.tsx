import Swal from "sweetalert2";
import { useAuthStore } from "../../hooks/useAuthStore";
import { useForm } from "../../hooks/useForm";
import "./LoginPage.css";
const loginFormFiles = {
  loginEmail: "",
  loginPassword: "",
};

const registerFormFiles = {
  registerName: "",
  registerEmail: "",
  registerPassword: "",
  registerConfirmPassword: "",
};

export const LoginPage = () => {
  const {
    loginEmail,
    loginPassword,
    onInputChange: onLoginInputChange,
  } = useForm(loginFormFiles);

  const {
    registerEmail,
    registerName,
    registerPassword,
    registerConfirmPassword,
    onInputChange: onRegisterInputChange,
  } = useForm(registerFormFiles);

  const {startLogin,startRegister}=useAuthStore()
  const loginSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    startLogin({email:loginEmail,password:loginPassword})
  };

  const registerSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if(registerPassword!==registerConfirmPassword) return Swal.fire('Error','Las contrase単as no coinciden','error')
    startRegister({email:registerEmail,password:registerPassword,name:registerName})
  }

  return (
    <div className="container login-container mx-auto ">
      <div className="grid grid-cols-12 gap-4 mx-auto">
        <div className="col-span-6 login-form-1">
          <h3>Ingreso</h3>
          <form onSubmit={loginSubmit}>
            <div className="mb-2">
              <input
                type="text"
                className="border-2 border-gray-300 rounded-md w-full p-2"
                placeholder="Correo"
                name="loginEmail"
                value={loginEmail}
                onChange={onLoginInputChange}
              />
            </div>
            <div className="mb-2">
              <input
                type="password"
                className="border-2 border-gray-300 rounded-md w-full p-2"
                placeholder="Contrase単a"
                name="loginPassword"
                value={loginPassword}
                onChange={onLoginInputChange}
              />
            </div>
            <div className="mb-2 flex items-center justify-center">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-span-6 login-form-2">
          <h3>Registro</h3>
          <form onSubmit={registerSubmit}>
            <div className="mb-2">
              <input
                type="text"
                className="border-2 border-gray-300 rounded-md w-full p-2"
                placeholder="Nombre"
                name="registerName"
                value={registerName}
                onChange={onRegisterInputChange}

              />
            </div>
            <div className="mb-2">
              <input
                type="email"
                className="border-2 border-gray-300 rounded-md w-full p-2"
                placeholder="Correo"
                name="registerEmail"
                value={registerEmail}
                onChange={onRegisterInputChange}
              />
            </div>
            <div className="mb-2">
              <input
                type="password"
                className="border-2 border-gray-300 rounded-md w-full p-2"
                placeholder="Contrase単a"
                name="registerPassword"
                value={registerPassword}
                onChange={onRegisterInputChange}

              />
            </div>

            <div className="mb-2">
              <input
                type="password"
                className="border-2 border-gray-300 rounded-md w-full p-2"
                placeholder="Repita la contrase単a"
                name="registerConfirmPassword"
                value={registerConfirmPassword}
                onChange={onRegisterInputChange}

              />
            </div>

            <div className="mb-2 flex items-center justify-center">
              <input type="submit" className="btnSubmit" value="Crear cuenta" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
