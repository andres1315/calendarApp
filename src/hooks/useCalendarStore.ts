import { EventCalendar, changeToActiveEvent, onAddNewEvent, onRemoveEvent, onUpdateEvent } from "../store/calendar/calendarSlice"
import { useAppDispatch, useAppSelector } from "../store/hooks"

export const useCalendarStore = ()=>{
  
  const dispatch = useAppDispatch()
  const {events, activeEvent} = useAppSelector(state => state.calendar)

  const setActieEvent = (event: EventCalendar) => {
    dispatch(changeToActiveEvent(event))
  }

  const startSavingEvent = async(calendarEvent) => {
    // todo legar al back

    // todo actualizar el store
    if (calendarEvent._id) {
      // actualizar
      dispatch(onUpdateEvent({...calendarEvent}))
    } else {
      // crear
      dispatch(onAddNewEvent({...calendarEvent, _id: new Date().getTime().toString()}))
    }
  }


  const deleteEvent=()=>{
    dispatch(onRemoveEvent())
  }
  
  return {
    events,
    activeEvent,
    setActieEvent,
    startSavingEvent,
    deleteEvent
  }
}