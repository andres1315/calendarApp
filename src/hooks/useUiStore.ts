import { useAppDispatch, useAppSelector } from "../store/hooks"
import { onCloseDateModal, onOpenDateModal } from "../store/ui/uiSlice"

export const useUiStore = () => {
  const {isDateModalOpen} = useAppSelector(state => state.ui)
  const dispatch = useAppDispatch()


  const onOpenModal = () => {
    dispatch(onOpenDateModal())
  }

  const onCloseModal = () => {
    dispatch(onCloseDateModal())
  } 

  return {
    //* props
    isDateModalOpen,
    //* methods
    onOpenModal,
    onCloseModal
  }
}