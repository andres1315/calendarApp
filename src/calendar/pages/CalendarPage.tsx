import { Calendar} from 'react-big-calendar'
import 'react-big-calendar/lib/css/react-big-calendar.css'
import {  NavbarCalendar } from "../components/Navbar"
import { localizer,getMessagesES } from '../../helpers'
import { useState } from 'react'
import { CalendarEvents } from '../components/CalendarEvents'
import { CalendarModal } from '../components/CalendarModal'
import { useUiStore } from '../../hooks/useUiStore'
import { useCalendarStore } from '../../hooks/useCalendarStore'
import { FabAddNew } from '../components/FabAddNew'
import { FaPlusCircle,FaRegTrashAlt } from "react-icons/fa";
import { FabDelete } from '../components/FabDelete'


const eventStyleGetter = (event: any, start: Date, end: Date, isSelected: boolean) => {
  return {
    className: 'custom-event-class',
    style: {
      backgroundColor: '#347cf7',
      opacity: 0.8,
      color: 'white',
    },
  };
}





export const CalendarPage = () => {
  const [lastView, setLastView] = useState<string>(localStorage.getItem('lastView') || 'week')
  const {onOpenModal} = useUiStore()
  const {events, activeEvent,setActieEvent} = useCalendarStore()


  const onDoubleClick = (e: any) => {
    onOpenModal()
  }
  
  const onSelect = (e: any) => {
    setActieEvent(e)
  }
  
  const onViewChange = (e: any) => {
    setLastView(e)
    localStorage.setItem('lastView',e)
  }
  return (
    <>
      <NavbarCalendar />  
      <Calendar
        culture='es'
        localizer={localizer}
        events={events}
        defaultView={lastView}
        startAccessor="start"
        endAccessor="end"
        style={{ height: 'calc(100vh - 65px)' }}
        messages={getMessagesES}
        eventPropGetter={eventStyleGetter}
        components={{
          event: CalendarEvents
        }}
        onDoubleClickEvent={onDoubleClick}
        onSelectEvent={onSelect}
        onView={onViewChange}
      />
      <CalendarModal />
      <FabAddNew className='animate__animated animate__pulse animate__infinite'>
        <FaPlusCircle size={30} />
      </FabAddNew>
      {
        activeEvent && (
          <FabDelete className='animate__animated animate__pulse animate__infinite'>
            <FaRegTrashAlt size={30} />
          </FabDelete>
        )
      }

      
    </>
  )
}
