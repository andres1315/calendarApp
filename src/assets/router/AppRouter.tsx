import { Route, Routes } from "react-router-dom"
import { LoginPage } from "../../auth/pages/LoginPage"
import { CalendarPage } from "../../calendar/pages/CalendarPage"

export const AppRouter = () => {
  const authStatus =  'unauthenticated'
  return (
    <Routes>
      {/* {authStatus === 'unauthenticated'
        ? <Route path='/auth/*' element={<LoginPage />} />
        : <Route path='/*' element={<CalendarPage />} />
      } */}
      
        <Route path='/auth/*' element={<LoginPage />} />
        <Route path='/*' element={<CalendarPage />} />
      
      <Route path='*' element={<div>404</div>} />
    </Routes>
  )
}
