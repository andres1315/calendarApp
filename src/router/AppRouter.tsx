import { Navigate, Route, Routes } from "react-router-dom"
import { LoginPage } from "../auth/pages/LoginPage"
import { CalendarPage } from "../calendar/pages/CalendarPage"
import { useAuthStore } from "../hooks/useAuthStore"
import { useEffect } from "react"

export const AppRouter = () => {
  const {checkToken,status} = useAuthStore()

  useEffect(() => {
    checkToken()
  },[])


  if (status === 'checking') return <h1>Checking...</h1>
  return (
    <Routes>
      {status === 'not-authenticated'
        ? (
          <>
            <Route path='/auth/*' element={<LoginPage />} />
            <Route path='*' element={<Navigate to='/auth/login'/>} />

          </>
        )
        :( 
          <>
            <Route path='/' element={<CalendarPage />} />
            <Route path='*' element={<Navigate to='/'/>} />

          </>
        )
      }
      

      
    </Routes>
  )
}
