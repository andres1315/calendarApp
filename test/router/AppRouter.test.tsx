import { render, screen } from "@testing-library/react"
import { AppRouter } from "../../src/router/AppRouter"
import { useAuthStore } from "../../src/hooks/useAuthStore"
import { be } from "date-fns/locale"
import { MemoryRouter } from "react-router-dom"
import { CalendarPage } from "../../src/calendar/pages/CalendarPage"
jest.mock('../../src/hooks/useAuthStore')
jest.mock('../../src/calendar/pages/CalendarPage',()=>(
  {
    CalendarPage: () => <div>CalendarPage</div>
  }
))
describe('pruebas en AppRouter', () => {

  const checkAuthToken = jest.fn()
  beforeEach(() => jest.clearAllMocks())
  test('debe de mostrar el mensaje "Checking..."', () => {
    useAuthStore.mockReturnValue({
      status: 'checking',
      checkToken: checkAuthToken
    })
    render(<AppRouter />)
    const text = screen.getByText('Checking...')
    expect(text).toBeTruthy()
    expect(checkAuthToken).toHaveBeenCalled()


  })
  
  test('debe de mostrar el componente <CalendarPage />', () => {
    useAuthStore.mockReturnValue({
      status: 'authenticated',
      checkToken: checkAuthToken
    })

    render(
    <MemoryRouter >
        <AppRouter />
    </MemoryRouter>)
    expect(screen.getByText('CalendarPage')).toBeTruthy()
    screen.debug()
  })
  
  test('debe de mostrar el componente <LoginPage />', () => {
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkToken: checkAuthToken
    })

    const {container} = render(<MemoryRouter>
        <AppRouter />
    </MemoryRouter>)
    expect(screen.getByText('Login')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })
})