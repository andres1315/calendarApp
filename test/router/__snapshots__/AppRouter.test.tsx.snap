// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en AppRouter debe de mostrar el componente <LoginPage /> 1`] = `
<div>
  <div
    class="container login-container mx-auto "
  >
    <div
      class="grid grid-cols-12 gap-4 mx-auto"
    >
      <div
        class="col-span-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="mb-2"
          >
            <input
              class="border-2 border-gray-300 rounded-md w-full p-2"
              name="loginEmail"
              placeholder="Correo"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-2"
          >
            <input
              class="border-2 border-gray-300 rounded-md w-full p-2"
              name="loginPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="mb-2 flex items-center justify-center"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-span-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="mb-2"
          >
            <input
              class="border-2 border-gray-300 rounded-md w-full p-2"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-2"
          >
            <input
              class="border-2 border-gray-300 rounded-md w-full p-2"
              name="registerEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="mb-2"
          >
            <input
              class="border-2 border-gray-300 rounded-md w-full p-2"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="mb-2"
          >
            <input
              class="border-2 border-gray-300 rounded-md w-full p-2"
              name="registerConfirmPassword"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="mb-2 flex items-center justify-center"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
